lib/jobs/run-graph.js

```diff
@@ -43,57 +47,97 @@ function runGraphJobFactory(
         this.parentGraphId = this.context.graphId;
         this.proxy = context.proxy;
+        this.graphPollInterval = this.options.graphPollInterval || 10000;
+
+        if(this.constructor.name === 'RunGraphJob') {
+            //only do these things for this class and not its subclasses
+            assert.string(this.options.graphName);
+            assert.object(this.options.graphOptions);
+            this.options.graphOptions.instanceId = this.graphId;
+            this.graphTarget = this.options.graphOptions.target;
+            this.domain = this.options.domain || Constants.Task.DefaultDomain;
+        }
     }
     util.inherits(RunGraphJob, BaseJob);
 
+
     /**
      * @memberOf RunGraphJob
      */
     RunGraphJob.prototype._run = function() {
         var self = this;
-
-        this._subscribeGraphFinished(function(status) {
-            if (status === Constants.Task.States.Succeeded) {
-                self._done();
-            } else {
-                self._done(new Error("Graph " + self.graphId + " failed with status " + status));
+        self.subscribeOwnGraph().then(function(ownGraph) {
+            self.subgraphPoll = self.setSubgraphPoll();
+            if (!ownGraph) {
+            return workflowTool.runGraph(
+                    self.graphTarget,
+                    self.options.graphName,
+                    self.options.graphOptions,
+                    self.domain,
+                    self.proxy,
+                    self.parentGraphId,
+                    self.taskId
+                );
             }
+        })
+        .catch(function(error) {
+           logger.error("Error starting graph for task.", {
+                    taskId: self.taskId,
+                    graphName: self.options.graphName,
+                    error: error
+                });
+            self._done(error);
         });
+    };
 
-        workflowTool.runGraph(
-            self.graphTarget,
-            self.options.graphName,
-            self.options.graphOptions,
-            self.domain,
-            self.proxy,
-            self.parentGraphId,
-            self.taskId
-        ).catch(function(error) {
-            logger.error("Error starting graph for task.", {
-                taskId: self.taskId,
-                graphName: self.options.graphName,
-                error: error
-            });
-            self._done(error);
+    RunGraphJob.prototype.finishWithState = function(subGraphStatus) {
+        if (subGraphStatus === Constants.Task.States.Succeeded) {
+            this._done();
+        } else if (_.contains(Constants.Task.FailedStates, subGraphStatus)) {
+            this._done(new Error(
+                "Graph " + this.graphId + " failed with status " + subGraphStatus)
+            );
+        }
+    };
+
+    RunGraphJob.prototype.subscribeOwnGraph = function() {
+        var self = this;
+        return waterline.graphobjects.findOneMongo({"context._parent.taskId": self.taskId})
```

Comment:
Can you call this behind the store layer, that we don't have mongo-specific code in the job code.